src/lib/device-logs/backends/loki.ts

```diff
@@ -0,0 +1,302 @@
+import * as _ from 'lodash';
+import * as grpc from 'grpc';
+import { EventEmitter } from 'events';
+
+import {
+	createInsecureCredentials,
+	createOrgIdMetadata,
+	Direction,
+	EntryAdapter,
+	PusherClient,
+	PushRequest,
+	PushResponse,
+	QuerierClient,
+	QueryRequest,
+	QueryResponse,
+	StreamAdapter,
+	TailRequest,
+	TailResponse,
+	Timestamp,
+} from 'loki-grpc-client';
+import { sbvrUtils } from '@resin/pinejs';
+import { LOKI_HOST, LOKI_PORT } from '../../../lib/config';
+import {
+	DeviceLog,
+	DeviceLogsBackend,
+	LogContext,
+	LogWriteContext,
+	Subscription,
+} from '../struct';
+import { captureException } from '../../../platform/errors';
+
+const { BadRequestError } = sbvrUtils;
+
+const MAX_RETRIES = 10;
+const MIN_BACKOFF = 100;
+const MAX_BACKOFF = 10 * 1000;
+const VERSION = 1;
+
+function sleep(ms: number) {
+	return new Promise((resolve) => setTimeout(resolve, ms));
+}
+
+function createTimestampFromDate(date = new Date()) {
+	const timestamp = new Timestamp();
+	timestamp.fromDate(date);
+	return timestamp;
+}
+
+function createTimestampFromMilliseconds(milliseconds: number) {
+	const MS_TO_NANOS = 1000000;
+	const seconds = Math.floor(milliseconds / 1000);
+	const nanos = (milliseconds - seconds * 1000) * MS_TO_NANOS;
+
+	const timestamp = new Timestamp();
+	timestamp.setSeconds(seconds);
+	timestamp.setNanos(nanos);
```

Comment:
Is it possible to pass this in the constructor, eg
```suggestion
	const timestamp = new Timestamp(seconds, nanos);
```
would be far nicer and seems sane enough as an option - I know we might be restricted by the library but if it's possible to do it'd be nice to

Reply:
Yeah its awkward, this is auto generated from the proto schema. I could give it a wrapper but that would be one more thing to check/maintain when updating the schema for new versions.

Reply:
If we don't have control over it then that's fair enough, I was just being hopeful :P